### DISCUSSION TITLE
How to generate ASTs instead of CSTs?

### DISCUSSION THREAD

**Discussion by brando90** [lark-parser/lark#1056] [Q&A] [answered]
**What is your question?**

I have simply type lambda calculus grammar. But when I use it and variations of it due to how parsers work I get a giant CST (compare to ASTs). I want to convert my CST to an AST with constructors Abs, Apply, Vars. Or directly from strings to ASTs. How does one do that?

**If you're having trouble with your code or grammar**
```
term : "λ" term ":" type "." term
    | "[" term term "]"
    | /[A-Za-z0-9_]+/
type : type "->" type  // "->"
    | /[A-Za-z0-9_]+/
// ignore tokens
IGNORE_TOKENS : "(" | ")"
%import common.WS
%ignore WS
%ignore IGNORE_TOKENS
// %import common.ESCAPED_STRING
// %import common.SIGNED_NUMBER
```

related: 
- https://cs.stackexchange.com/questions/146612/how-does-one-parse-a-string-into-an-ast-abstract-syntax-tree-directly-instead
- https://github.com/lark-parser/lark/issues/1053

**Comment by brando90** [comment, marked as answer]
from erezsh:

This might help: https://lark-parser.readthedocs.io/en/latest/examples/advanced/create_ast.html

Also useful to have a good grasp of the built-in tree construction option: https://lark-parser.readthedocs.io/en/latest/tree_construction.html

from: https://github.com/lark-parser/lark/issues/1053

**Reply by brando90** [reply]
> This might help: https://lark-parser.readthedocs.io/en/latest/examples/advanced/create_ast.html
> 
> Also useful to have a good grasp of the built-in tree construction option: https://lark-parser.readthedocs.io/en/latest/tree_construction.html

btw @erezsh, out of curiosity is CST -> AST a very standard PL (programming language) thing?

**Reply by erezsh** [reply]
Yeah, I'd say so. ASTs are the standard representation of code in compilers, and most parsers generate CSTs. Then a conversion is unavoidable.
